geopandas/geodataframe.py

```diff
@@ -1384,18 +1415,34 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
         result = super().apply(
             func, axis=axis, raw=raw, result_type=result_type, args=args, **kwargs
         )
-        if (
-            isinstance(result, GeoDataFrame)
-            and self._geometry_column_name in result.columns
-            and any(isinstance(t, GeometryDtype) for t in result.dtypes)
-        ):
-            if self.crs is not None and result.crs is None:
-                result.set_crs(self.crs, inplace=True)
+
+        if self._geometry_column_name in result.columns:
+            # Reconstruct gdf if it was lost by apply
+            if isinstance(result, GeoDataFrame) is False:
+                # axis=1 apply will split GeometryDType to object, try and cast back
+                try:
+                    _ensure_geometry(result[self._geometry_column_name], crs=self.crs)
```

Comment:
This unfortunately is, `test_crs.py::test_apply_geodataframe` fails without it for the axis=1 case:
```python
df = GeoDataFrame({"col1": [0, 1]}, geometry=self.geoms, crs=27700)
result = df.apply(lambda row: row, axis=1)
assert result.crs == 27700
```
this fails because `result` is a dataframe now. I will have a look and see if there's a simpler solution though.

Reply:
Ah, yes, I see. We _could_ maybe also move this logic of trying to convert a geometry column in the `_geodataframe_constructor_with_fallback`. But until we discover more cases where that might be useful, it's fine to keep this in `apply`.